### PR TITLE
Autodan

### PR DESCRIPTION
**Pull request by erickgalinkin**
Add AutoDAN implementation (and skeleton of interactive mode, which belongs in a different branch)

### CODE REVIEW

**Comment by erickgalinkin** [comment]
I have read the DCO Document and I hereby sign the DCO

**Comment by leondz** [comment]
recheck

File: garak/cli.py [hunk 1]
```diff
@@ -262,6 +281,16 @@ def print_plugins(prefix, color):
             )
         generator.generations = _config.args.generations
 
+        if _config.args.generate_autodan:
+            from garak.resources.autodan import autodan_generate
+            try:
+                prompt = _config.probe_options["prompt"]
+                target = _config.probe_options["target"]
+            except Exception as e:
+                print("AutoDAN generation requires --prompt_options with a .json containing a `prompt` and `target` "
```
Comment (leondz) [inline garak/cli.py:290]: --probe_options?

File: garak/generators/__init__.py
Comment (leondz) [inline garak/generators/__init__.py:1]: I guess this is from cli.py. I wonder where's best to put it? Definitely not cli.py; maybe _plugins.py? Maybe _plugins.py should delegate loading stuff out to package ``__init__``s implemneted like this - then the generator-specific code is in the generator package, the snozzberries taste like snozzberries, ...

Comment (leondz) [inline garak/generators/__init__.py:1]: May need SPDX headers, see lines 3+4 in https://github.com/leondz/garak/blob/8ed28204fb74e2be857847426b4b04af17b9d6a1/garak/_plugins.py

File: garak/interactive.py
Comment (leondz) [inline garak/interactive.py:1]: can there be a header w/
* shebang /usr/bin/env python3
* blank line
* SPDX headers for copyright (not partial, just copyright)

Comment (leondz) [inline garak/interactive.py:1]: can garak imports be grouped after system ones?

Comment (leondz) [inline garak/interactive.py:1]: a refactor is now possible w/ `command.py` but this is not high priority

File: garak/interactive.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        self.add_settable(cmd2.Settable("generator", str, "Generator settings path", self))
+        self.add_settable(cmd2.Settable("eval_threshold", float, "Evaluation threshold for success", self))
+        # Set default parameter values
+        self.target_type = ""
+        self.target_model = ""
+        self.probe = ""
+        self.detector = ""
+        self.generator = ""
+        self.eval_threshold = 0.5
+        # Disable shell, script, alias, and edit commands
```
Comment (leondz) [inline garak/interactive.py:142]: love this section

File: garak/probes/dan.py [hunk 3]
```diff
@@ -371,3 +372,25 @@ def probe(self, generator):
...
         return super().probe(generator)
+
+
+class AutoDANProbe(Probe):
+    """Probe leveraging generated AutoDAN prompts"""
+    bcp47 = "en"
+    recommended_detector = ["mitigation.MitigationBypass", "dan.DAN"]
+    tags = ["avid-effect:security:S0403"]
+    goal = "disregard the system prompt"
+
```
Comment (leondz) [inline garak/probes/dan.py:383]: maybe add a `uri = ` ... wait there are two autodans now?? `https://arxiv.org/abs/2310.04451` / `https://arxiv.org/abs/2310.15140`

File: garak/probes/transferableattacks.py
Comment (leondz) [inline garak/probes/transferableattacks.py:1]: needs an SPDX two-liner

File: garak/resources/autodan/__init__.py
Comment (leondz) [inline garak/resources/autodan/__init__.py:1]: SPDX sprinkles here too ✨

File: garak/resources/autodan/autodan.py
Comment (leondz) [inline garak/resources/autodan/autodan.py:1]: probably needs a copy of the original licence for autodan and copyright statement, so it's clear it's not our code - what license was it?

File: garak/resources/autodan/genetic.py
Comment (leondz) [inline garak/resources/autodan/genetic.py:1]: this one needs a license at the top. also - can the garak imports be grouped at the bottom? the rough convention has been stdlib-y imports, pip imports, garak imports

File: garak/resources/autodan/model_utils.py
Comment (leondz) [inline garak/resources/autodan/model_utils.py:1]: this looks strong. i guess it just needs a license. oh, and the required libs (+versions) shoudl also be in the 'pyproject.toml' and `requirements.txt`

File: garak/cli.py (hunk 1 above)
Comment (erickgalinkin) [reply garak/cli.py:290]: Good catch!

File: garak/generators/__init__.py
Comment (erickgalinkin) [reply garak/generators/__init__.py:1]: Yeah, I think this is worth a longer conversation and implementing a consistent pattern throughout.

File: garak/interactive.py
Comment (erickgalinkin) [reply garak/interactive.py:1]: Will pull this into a different branch focused specifically on the interactive mode.

File: garak/resources/autodan/model_utils.py
Comment (erickgalinkin) [reply garak/resources/autodan/model_utils.py:1]: Good catch on `pyproject.toml` and `requirements.txt`!

File: garak/resources/autodan/autodan.py
Comment (erickgalinkin) [reply garak/resources/autodan/autodan.py:1]: MIT license -- We've modified it, and I ref the original license in __init__.py.

**Pull request merged**